### PR TITLE
Update Kafka Client to 2.8.0

### PR DESCRIPTION
**Pull request by eolivelli**
Summary of changes:
- Bump client dependency from 2.1.x to 2.8.x
- The Java Model for all the Request/Response changed to a Fluent/Builder style, this requires many changes
- NotLeaderForPartitionException -> NotLeaderOrFollowerException
- Copy new version of MockTime from Kafka codebase
- ListOffsetRequest -> ListOffsetsRequest, ListOffsetResponse -> ListOffsetsResponse
- LeaveGroup -> now support multiple members

### CODE REVIEW

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaCommandDecoder.java [hunk 1]
```diff
@@ -144,13 +144,13 @@ protected KafkaHeaderAndRequest byteBufToRequest(ByteBuf msg,
         ByteBuffer nio = msg.nioBuffer();
         RequestHeader header = RequestHeader.parse(nio);
         if (isUnsupportedApiVersionsRequest(header)) {
-            ApiVersionsRequest apiVersionsRequest = new ApiVersionsRequest((short) 0, header.apiVersion());
+            ApiVersionsRequestData data = new ApiVersionsRequestData();
+            ApiVersionsRequest apiVersionsRequest = new ApiVersionsRequest(data, header.apiVersion());
```
Comment (Demogorgon314) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaCommandDecoder.java:148]: Use `new ApiVersionsRequest.Builder(header.apiVersion()).build();`?

Comment (eolivelli) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaCommandDecoder.java:148]: makes sense, updated ! thanks

**Comment by Demogorgon314** [comment]
Overall, LGTM. I will wait for the unit's test to pass.

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/CorrelationIdGenerator.java [hunk 2]
```diff
@@ -17,7 +17,7 @@ public class CorrelationIdGenerator {
 
     private static final int MAX_RESERVED_CORRELATION_ID = Integer.MAX_VALUE;
     private static final int MIN_RESERVED_CORRELATION_ID = MAX_RESERVED_CORRELATION_ID - 7;
-    private int correlationId = 0;
+    private int correlationId = 999;
```
[the remark this answers is not in the document]
Comment (eolivelli) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/CorrelationIdGenerator.java:20]: It is a left over. I had to debug the wire protocol and having 0 was confusing.
I have reverted it to 0, there is no reason to use 999.
good catch

**Comment by eolivelli** [comment]
@Demogorgon314 @Demogorgon314 this patch is finally ready for review.
All the tests are passing locally on my laptop and Codacy is happy

**Comment by eolivelli** [comment]
CI passed

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java [hunk 3]
```diff
@@ -1750,11 +1800,20 @@ protected void handleLeaveGroupRequest(KafkaHeaderAndRequest leaveGroup,
...
         LeaveGroupRequest request = (LeaveGroupRequest) leaveGroup.getRequest();
+        LeaveGroupRequestData data = request.data();
+        Set<String> members = new HashSet<>();
+        if (!data.memberId().isEmpty()) {
+            // old clients
+            members.add(data.memberId());
+        }
+        data.members().forEach(memberIdentity -> {
+            members.add(memberIdentity.memberId());
+        });
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java:1811]: We can use functional style here

```java
        Set<String> members = data.members().stream()
                .map(LeaveGroupRequestData.MemberIdentity::memberId)
                .collect(Collectors.toSet());
        if (!data.memberId().isEmpty()) {
            // old clients
            members.add(data.memberId());
        }
```

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java [hunk 4]
```diff
@@ -1784,17 +1843,20 @@ protected void handleDeleteGroupsRequest(KafkaHeaderAndRequest deleteGroups,
...
 
     @Override
     protected void handleSaslAuthenticate(KafkaHeaderAndRequest saslAuthenticate,
                                           CompletableFuture<AbstractResponse> resultFuture) {
-        resultFuture.complete(new SaslAuthenticateResponse(Errors.ILLEGAL_SASL_STATE,
-                "SaslAuthenticate request received after successful authentication"));
+        SaslAuthenticateResponseData data = new SaslAuthenticateResponseData();
+        data.setErrorCode(Errors.ILLEGAL_SASL_STATE.code());
+        data.setErrorMessage("SaslAuthenticate request received after successful authentication");
+        resultFuture.complete(new SaslAuthenticateResponse(data));
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java:1859]:
```suggestion
        resultFuture.complete(new SaslAuthenticateResponse(
                new SaslAuthenticateResponseData()
                        .setErrorCode(Errors.ILLEGAL_SASL_STATE.code())
                        .setErrorMessage("SaslAuthenticate request received after successful authentication")));
```

I think Kafka's new data structure is designed for chained calls above. The *data* is mutable while the *response* that owns the data is immutable.

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java [hunk 5]
```diff
@@ -1894,56 +1956,95 @@ protected void handleAlterConfigs(KafkaHeaderAndRequest describeConfigs,
...
+                            .resources()
+                            .stream()
+                            .filter(r->r.resourceName().equals(configResource.name())
+                                    && r.resourceType() == configResource.type().id())
+                            .findAny()
+                            .map(__ -> new HashSet<>());
+                    resourceToConfigNames.put(configResource, configNames);
+                });
+
+                adminManager.describeConfigsAsync(resourceToConfigNames, namespacePrefix
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java:2016]:
```suggestion
                adminManager.describeConfigsAsync(authorizedResources.stream()
                                .collect(Collectors.toMap(
                                        configResource -> configResource,
                                        configResource -> data.resources().stream()
                                                .filter(r -> r.resourceName().equals(configResource.name())
                                                        && r.resourceType() == configResource.type().id())
                                                .findAny()
                                                .map(__ -> new HashSet<>())
                                )),
                        namespacePrefix
                ).thenApply(configResourceConfigMap -> {
```

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaApisTest.java [hunk 6]
```diff
@@ -210,15 +196,39 @@ public void testOffsetCommitWithInvalidPartition() throws Exception {
...
+                        .filter(t->t.name().equals(topicName))
+                        .findFirst()
+                        .get()
+                        .partitions()
+                        .stream()
+                        .filter(p -> p.partitionIndex() == topicPartition1.partition())
+                        .findFirst()
+                        .get()
+                        .errorCode(),
             Errors.UNKNOWN_TOPIC_OR_PARTITION);
```
Comment (BewareMyPower) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaApisTest.java:212]:
```suggestion
            Errors.UNKNOWN_TOPIC_OR_PARTITION.code());
```

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaApisTest.java [hunk 7]
```diff
@@ -210,15 +196,39 @@ public void testOffsetCommitWithInvalidPartition() throws Exception {
...
+                                .filter(t->t.name().equals(topicName))
+                                .findFirst()
+                                .get()
+                                .partitions()
+                                .stream()
+                                .filter(p -> p.partitionIndex() == topicPartition2.partition())
+                                .findFirst()
+                                .get()
+                                .errorCode(),
             Errors.UNKNOWN_TOPIC_OR_PARTITION);
```
Comment (BewareMyPower) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaApisTest.java:232]:
```suggestion
            Errors.UNKNOWN_TOPIC_OR_PARTITION.code());
```

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaApisTest.java [hunk 8]
```diff
@@ -210,15 +196,39 @@ public void testOffsetCommitWithInvalidPartition() throws Exception {
...
+                        .errorCode(),
             Errors.UNKNOWN_TOPIC_OR_PARTITION);
 
         // invalid partition id 1.
         CompletableFuture<AbstractResponse> invalidResponse2 = new CompletableFuture<>();
         checkInvalidPartition(invalidResponse2, topicName, 1);
         TopicPartition topicPartition2 = new TopicPartition(topicName, 1);
         AbstractResponse response2 = invalidResponse2.get();
-        assertEquals(((OffsetCommitResponse) response2).responseData().get(topicPartition2),
+                assertEquals(((OffsetCommitResponse) response2)
```
Comment (BewareMyPower) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaApisTest.java:219]:
```suggestion
        assertEquals(((OffsetCommitResponse) response2)
```

Adjust indent.

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaAuthorizationTestBase.java [hunk 9]
```diff
@@ -680,6 +680,7 @@ public void testDescribeConfigFailed() throws PulsarAdminException {
         try {
             describeConfigsResult.all().get();
         } catch (Exception ex) {
+            log.error("Error", ex);
```
Comment (BewareMyPower) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaAuthorizationTestBase.java:683]:
```suggestion
```

Remove this line because the exception is expected. Or just use `log.info("Error: {}", ex.getMessage()`. BTW, I found here we should add a `fail()` in the `try` block.

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaListenerNameTest.java [hunk 10]
```diff
@@ -117,11 +117,13 @@ public void testMetadataRequestForMultiListeners() throws Exception {
...
                     final PartitionMetadata partitionMetadata = partitionMetadataList.get(i);
-                    Assert.assertEquals(partitionMetadata.error(), Errors.NONE);
-                    Assert.assertEquals(partitionMetadata.leader().host() + ":" + partitionMetadata.leader().port(),
+                    Assert.assertEquals(partitionMetadata.error, Errors.NONE);
+                    Node leader = metadataResponse.brokersById().get(partitionMetadata.leaderId.get());
+                    Assert.assertEquals(leader.host() + ":" + leader.port(),
                             expectedAddress);
                 }
             } catch (Exception e) {
+                log.error("Error", e);
```
Comment (BewareMyPower) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaListenerNameTest.java:126]:
```suggestion
```

Don't print the whole exception stack if it's expected.

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaListenerNameTest.java [hunk 11]
```diff
@@ -63,7 +63,7 @@ protected void cleanup() throws Exception {
         // Clean up in the test method
     }
 
-    @Test(timeOut = 30000)
+    @Test
```
Comment (BewareMyPower) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaListenerNameTest.java:66]: Why did you remove the timeout?

**Comment by BewareMyPower** [comment]
My review is done, PTAL.

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaAuthorizationTestBase.java (hunk 9 above)
Comment (eolivelli) [reply tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaAuthorizationTestBase.java]: fixed

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaListenerNameTest.java (hunk 11 above)
Comment (eolivelli) [reply tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaListenerNameTest.java]: it was only to use the debugger :-) 

restored. good catch

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaListenerNameTest.java (hunk 10 above)
Comment (eolivelli) [reply tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaListenerNameTest.java]: done

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java (hunk 3 above)
Comment (eolivelli) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java]: sure

**Comment by eolivelli** [comment]
@BewareMyPower what about the "snyk" failures ? before my last commit it was passing

**Comment by eolivelli** [comment]
@BewareMyPower what about the "snyk" failures ? before my last commit it was passing

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java [hunk 12]
```diff
@@ -1661,29 +1703,31 @@ protected void handleJoinGroupRequest(KafkaHeaderAndRequest joinGroup,
...
+                data.sessionTimeoutMs(),
+                data.protocolType(),
             protocols
         ).thenAccept(joinGroupResult -> {
 
-            Map<String, ByteBuffer> members = new HashMap<>();
+            Map<String, byte[]> members = new HashMap<>();
             joinGroupResult.getMembers().forEach((memberId, protocol) ->
-                members.put(memberId, ByteBuffer.wrap(protocol)));
+                members.put(memberId, protocol));
```
Comment (Demogorgon314) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java:1724]:
```java
Map<String, byte[]> members = new HashMap<>(joinGroupResult.getMembers());
```

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java [hunk 13]
```diff
@@ -1894,56 +1955,91 @@ protected void handleAlterConfigs(KafkaHeaderAndRequest describeConfigs,
...
                                          CompletableFuture<AbstractResponse> resultFuture) {
         checkArgument(describeConfigs.getRequest() instanceof DescribeConfigsRequest);
         DescribeConfigsRequest request = (DescribeConfigsRequest) describeConfigs.getRequest();
+        DescribeConfigsRequestData data = request.data();
 
-        if (request.resources().isEmpty()) {
-            resultFuture.complete(new DescribeConfigsResponse(0, Maps.newHashMap()));
+        if (data.resources().isEmpty()) {
+            DescribeConfigsResponseData responseData = new DescribeConfigsResponseData();
+            resultFuture.complete(new DescribeConfigsResponse(responseData));
```
Comment (Demogorgon314) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java:1988]:
```suggestion
            resultFuture.complete(new DescribeConfigsResponse(new DescribeConfigsResponseData()));
```

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java [hunk 14]
```diff
@@ -1999,18 +2096,24 @@ protected void handleDescribeConfigs(KafkaHeaderAndRequest describeConfigs,
...
-                request.transactionalId(), request.transactionTimeoutMs(), Optional.empty(),
-                (resp) -> response.complete(
-                        new InitProducerIdResponse(0, resp.getError(), resp.getProducerId(),
-                                resp.getProducerEpoch())));
+                data.transactionalId(), data.transactionTimeoutMs(), Optional.empty(),
+                (resp) -> {
+                    InitProducerIdResponseData responseData = new InitProducerIdResponseData();
+                    responseData.setErrorCode(resp.getError().code());
+                    responseData.setProducerId(resp.getProducerId());
+                    responseData.setProducerEpoch(resp.getProducerEpoch());
```
Comment (Demogorgon314) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java:2107]:
```suggestion
                    InitProducerIdResponseData responseData = new InitProducerIdResponseData()
                            .setProducerId(resp.getProducerId())
                            .setProducerEpoch(resp.getProducerEpoch())
                            .setErrorCode(resp.getError().code());
```

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupCoordinator.java [hunk 15]
```diff
@@ -555,20 +558,24 @@ private void doSyncGroup(GroupMetadata group,
...
         String groupId,
-        String memberId
+        Set<String> members
     ) {
         return validateGroupStatus(groupId, ApiKeys.LEAVE_GROUP).map(CompletableFuture::completedFuture
         ).orElseGet(() -> groupManager.getGroup(groupId).map(group -> group.inLock(() -> {
-            if (group.is(Dead) || !group.has(memberId)) {
+            if (group.is(Dead)) {
+                return CompletableFuture.completedFuture(Errors.UNKNOWN_MEMBER_ID);
+            } else if (!members.stream().allMatch((memberId) -> group.has(memberId))) {
```
Comment (Demogorgon314) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupCoordinator.java:567]:
```suggestion
            } else if (!members.stream().allMatch(group::has)) {
```
Lambda can be replaced with method reference

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupCoordinator.java [hunk 16]
```diff
@@ -555,20 +558,24 @@ private void doSyncGroup(GroupMetadata group,
...
     public CompletableFuture<Errors> handleLeaveGroup(
         String groupId,
-        String memberId
+        Set<String> members
     ) {
         return validateGroupStatus(groupId, ApiKeys.LEAVE_GROUP).map(CompletableFuture::completedFuture
         ).orElseGet(() -> groupManager.getGroup(groupId).map(group -> group.inLock(() -> {
-            if (group.is(Dead) || !group.has(memberId)) {
+            if (group.is(Dead)) {
+                return CompletableFuture.completedFuture(Errors.UNKNOWN_MEMBER_ID);
```
Comment (Demogorgon314) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupCoordinator.java:566]: The error should be `Errors.COORDINATOR_NOT_AVAILABLE`?

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/utils/KafkaResponseUtils.java [hunk 17]
```diff
@@ -52,153 +71,303 @@
...
-        return new ApiVersionsResponse(0, errors, Collections.emptyList());
+        ApiVersionsResponseData data = new ApiVersionsResponseData();
+        data.setErrorCode(errors.code());
+        return new ApiVersionsResponse(data);
     }
 
     public static CreatePartitionsResponse newCreatePartitions(Map<String, ApiError> topicToErrors) {
-        return new CreatePartitionsResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, topicToErrors);
+        CreatePartitionsResponseData data = new CreatePartitionsResponseData();
+        data.setThrottleTimeMs(AbstractResponse.DEFAULT_THROTTLE_TIME);
```
Comment (Demogorgon314) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/utils/KafkaResponseUtils.java:92]:
```suggestion
        CreatePartitionsResponseData data = new CreatePartitionsResponseData()
                .setThrottleTimeMs(AbstractResponse.DEFAULT_THROTTLE_TIME);
```

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/utils/KafkaResponseUtils.java [hunk 18]
```diff
@@ -52,153 +71,303 @@
...
+                    .setMinVersion(apiVersion.minVersion)
+                    .setMaxVersion(apiVersion.maxVersion));
+        });
+        return new ApiVersionsResponse(data);
     }
 
     public static ApiVersionsResponse newApiVersions(Errors errors) {
-        return new ApiVersionsResponse(0, errors, Collections.emptyList());
+        ApiVersionsResponseData data = new ApiVersionsResponseData();
+        data.setErrorCode(errors.code());
```
Comment (Demogorgon314) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/utils/KafkaResponseUtils.java:86]:
```suggestion
        ApiVersionsResponseData data = new ApiVersionsResponseData()
                .setErrorCode(errors.code());
```

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/utils/KafkaResponseUtils.java [hunk 19]
```diff
@@ -52,153 +71,303 @@
...
     }
 
     public static CreateTopicsResponse newCreateTopics(Map<String, ApiError> errorMap) {
-        return new CreateTopicsResponse(errorMap);
+        CreateTopicsResponseData data = new CreateTopicsResponseData();
+        errorMap.forEach((topic, errors) -> {
+            data.topics().add(new CreateTopicsResponseData.CreatableTopicResult()
+                            .setName(topic)
+                            .setErrorMessage(errors.messageWithFallback())
+                    .setErrorCode(errors.error().code()));
```
Comment (Demogorgon314) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/utils/KafkaResponseUtils.java:108]:
```suggestion
            data.topics().add(new CreateTopicsResponseData.CreatableTopicResult()
                    .setName(topic)
                    .setErrorMessage(errors.messageWithFallback())
                    .setErrorCode(errors.error().code()));
```

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaApisTest.java [hunk 20]
```diff
@@ -869,18 +891,20 @@ public void testGetOffsetsForUnknownTopic() throws Exception {
...
             .handleListOffsetRequest(request, responseFuture);
 
         AbstractResponse response = responseFuture.get();
-        ListOffsetResponse listOffsetResponse = (ListOffsetResponse) response;
-        assertEquals(listOffsetResponse.responseData().get(tp).error,
+        ListOffsetsResponse listOffsetsResponse = (ListOffsetsResponse) response;
+        ListOffsetsResponseData.ListOffsetsPartitionResponse listOffsetsPartitionResponse =
+                getListOffsetsPartitionResponse(tp, listOffsetsResponse.data());
+        assertEquals(listOffsetsPartitionResponse.errorCode(),
             Errors.UNKNOWN_TOPIC_OR_PARTITION);
```
Comment (Demogorgon314) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaApisTest.java:908]:
```suggestion
            Errors.UNKNOWN_TOPIC_OR_PARTITION.code());
```

**Review by Demogorgon314 — COMMENTED** [review]
LGTM. Just left some small comments.

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaAuthorizationTestBase.java (hunk 9 above)
Comment (eolivelli) [reply tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaAuthorizationTestBase.java]: @BewareMyPower I didn't add "fail()" for this PR, because with "fail()" actually the test will fail.
We can address this problem in the test in a follow up work.
it is not strictly related to this patch

**Review by eolivelli — COMMENTED** [review]
I have addresses your comments.
I hope I haven't missed any of them

thank you @Demogorgon314 @BewareMyPower

**Comment by eolivelli** [comment]
The tests that failed on CI pass on my laptop

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupCoordinator.java (hunk 16 above)
Comment (Demogorgon314) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupCoordinator.java:566]: Sure, I just found the Kafka 2.8 code is using `Errors.COORDINATOR_NOT_AVAILABLE`. We can change it in next PR if necessary.

https://github.com/apache/kafka/blob/2.8/core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala#L470-L471

**Review by Demogorgon314 — DISMISSED** [review]

**Review by BewareMyPower — APPROVED** [review]

**Comment by eolivelli** [comment]
@BewareMyPower is the failed test a flaky test ?
it doesn't fail locally for me

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupCoordinator.java (hunk 16 above)
Comment (eolivelli) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupCoordinator.java:566]: let me follow up with another patch on this

**Comment by eolivelli** [comment]
do you think that this patch is introducing flakyness ?
those tests don't fail locally

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupCoordinator.java (hunk 16 above)
Comment (eolivelli) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupCoordinator.java]: @Demogorgon314 finally I updated this value in this PR

**Comment by eolivelli** [comment]
I have rebased on master

**Comment by eolivelli** [comment]
> @eolivelli I can reproduce this failed test in my locally.

I will run the test many times. thanks for your feedback.
The test is pretty straight forward

**Comment by eolivelli** [comment]
@Demogorgon314 I have fixed the test.
it seems that the 2.8.0 client eagerly pre-fetches the messages and so it is able to pre-fetch the records before we delete the partition.

**Comment by eolivelli** [comment]
@BewareMyPower @Demogorgon314 can you please restart CI ?
I believe that this time we hit a flaky test.

**Comment by eolivelli** [comment]
the same test did not fail here 
https://github.com/streamnative/kop/pull/1606

**Comment by eolivelli** [comment]
All tests passed.

**Review by Demogorgon314 — APPROVED** [review]
